[expo-updates][Android] DatabaseLauncher called twice on same instance causing AssertionError when EmbeddedLoader fails in LoaderTask

### Minimal reproducible example

https://github.com/louveshh/minimal-reproducible-example-expo-updates

### Steps to reproduce

1. Build Android release APK with embedded update from the minimal reproducible example repository
2. Install the APK on an Android device
3. Clear app data: `adb shell pm clear com.yourcompany.expoupdatesbug`
4. Launch the app (cold start)
5. If the embedded loader encounters an error during startup (e.g., corrupted assets, network issues, or timing conditions), the app crashes

**Platform**: Android
**Environment**: Standalone app (release build)
**Package Manager**: npm

**Expected**: App should launch successfully and handle embedded update loading gracefully, even if `EmbeddedLoader.load()` throws an exception. The `DatabaseLauncher` should either prevent duplicate launches (idempotent) or create a new instance.

**Actual**: App crashes with `AssertionError: DatabaseLauncher has already started. Create a new instance in order to launch a new version.` at `expo.modules.updates.launcher.DatabaseLauncher.launch (DatabaseLauncher.kt:73)`.

**Root Cause**: In `LoaderTask.launchFallbackUpdateFromDisk()`, if `EmbeddedLoader.load()` throws an exception after `launcher.launch(database)` is called at line 302, the catch block at line 305 calls `launch()` again on the same instance, violating the assertion that prevents multiple launches.

**Stack Trace**:
```
Exception java.lang.AssertionError: DatabaseLauncher has already started. Create a new instance in order to launch a new version.
  at expo.modules.updates.launcher.DatabaseLauncher.launch (DatabaseLauncher.kt:73)
  at expo.modules.updates.loader.LoaderTask.launchFallbackUpdateFromDisk (LoaderTask.kt:305)
  at expo.modules.updates.loader.LoaderTask.start (LoaderTask.kt:170)
  at expo.modules.updates.procedures.StartupProcedure.run (StartupProcedure.kt:207)
  at expo.modules.updates.procedures.StateMachineSerialExecutorQueue.executeProcedure (StateMachineSerialExecutorQueue.kt:43)
  at expo.modules.updates.procedures.StateMachineSerialExecutorQueue.access$executeProcedure (StateMachineSerialExecutorQueue.kt:19)
  at expo.modules.updates.procedures.StateMachineSerialExecutorQueue$1.invokeSuspend (StateMachineSerialExecutorQueue.kt:36)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (ContinuationImpl.kt:33)
  at kotlinx.coroutines.DispatchedTask.run (DispatchedTask.kt:100)
  at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run (LimitedDispatcher.kt:113)
  at kotlinx.coroutines.scheduling.TaskImpl.run (Tasks.kt:89)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely (CoroutineScheduler.kt:586)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask (CoroutineScheduler.kt:820)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker (CoroutineScheduler.kt:717)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run (CoroutineScheduler.kt:704)
```

**Affected Devices**:
- POCO garnet - Qualcomm SM7435, 11,827 MB RAM
- Samsung dm3q - QTI SM8550, 11,582 MB RAM
- Samsung e3q - QTI SM8650, 12,288 MB RAM (11,624 MB available)
- Samsung pa1q - QTI SM8750, 12,288 MB RAM (11,654 MB available)
- Samsung pa3q - QTI SM8750, 12,288 MB RAM (11,654 MB available)
- Unnecto Bolt10 - Spreadtrum SC9863A, 4,096 MB RAM (4,007 MB available)

The issue affects devices across different price ranges and chipset manufacturers, suggesting it's not hardware-specific but related to the expo-updates library behavior.


### Environment

```text
expo-env-info 2.0.7 environment info:
    System:
      OS: macOS 26.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 24.8.0 - ~/.nvm/versions/node/v24.8.0/bin/node
      Yarn: 1.22.22 - /opt/homebrew/bin/yarn
      npm: 11.6.0 - ~/.nvm/versions/node/v24.8.0/bin/npm
      Watchman: 2025.09.15.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.16.2 - /Users/lukaszbaran/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 25.1, iOS 26.1, macOS 26.1, tvOS 26.1, visionOS 26.1, watchOS 26.1
    IDEs:
      Android Studio: 2025.1 AI-251.26094.121.2513.14007798
      Xcode: 26.1.1/17B100 - /usr/bin/xcodebuild
    npmPackages:
      expo: 54.0.25 => 54.0.25 
      expo-updates: ^29.0.12 => 29.0.12 
      react: 19.1.0 => 19.1.0 
      react-native: 0.81.5 => 0.81.5 
    npmGlobalPackages:
      eas-cli: 16.27.0
    Expo Workflow: managed
```

### Expo Doctor Diagnostics

```text
Running 17 checks on your project...
16/17 checks passed. 1 checks failed. Possible issues detected:

✖ Check that packages match versions required by installed Expo SDK

⚠️ Minor version mismatches
package                      expected  found   
@stripe/stripe-react-native  0.50.3    0.54.1  

1 package out of date.

Note: The Stripe version mismatch is intentional as it's part of the ProGuard configuration setup and doesn't affect the bug reproduction.
```

--- a/packages/expo-updates/CHANGELOG.md
+++ b/packages/expo-updates/CHANGELOG.md
@@ -11,7 +11,7 @@
 
 ### 🐛 Bug fixes
 
-- [Android] Fix issue where launch is called twice on the same database instance.
+- [Android] Fix issue where launch is called twice on the same database instance. ([#41152](https://github.com/expo/expo/pull/41152) by [@alanjhughes](https://github.com/alanjhughes))
 
 ### 💡 Others
 

--- a/packages/expo-updates/CHANGELOG.md
+++ b/packages/expo-updates/CHANGELOG.md
@@ -11,6 +11,8 @@
 
 ### 🐛 Bug fixes
 
+- [Android] Fix issue where launch is called twice on the same database instance.
+
 ### 💡 Others
 
 - [Android] Migrated from `kotlinOptions` to `compilerOptions` DSL. ([#39794](https://github.com/expo/expo/pull/39794) by [@huextrat](https://github.com/huextrat))

--- a/packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt
+++ b/packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt
@@ -296,14 +296,13 @@ class LoaderTask(
       ) {
         try {
           val embeddedLoader = EmbeddedLoader(context, configuration, logger, database, directory)
-          val result = embeddedLoader.load { updateResponse ->
+          embeddedLoader.load { _ ->
             Loader.OnUpdateResponseLoadedResult(shouldDownloadManifestIfPresentInResponse = true)
           }
-          launcher.launch(database)
         } catch (e: Exception) {
           logger.error("Unexpected error copying embedded update", e, UpdatesErrorCode.Unknown)
-          launcher.launch(database)
         }
+        launcher.launch(database)
       } else {
         launcher.launch(database)
       }

